### PR TITLE
Monica Nguyen pull request

### PR DESCRIPTION
**Pull request by Monica20030707**
For Task 1

### CODE REVIEW

File: src/main/java/StringManipulation.java [hunk 1]
```diff
@@ -1,32 +1,119 @@
...
+
+        return result;
     }
 
     @Override
-    public String[] getSubStrings(int startWord, int endWord) {
-        return null;
+    public String[] getSubStrings(int startWord, int endWord){
+        // check if the beginning word and end word less than 0 and conflict with each other
+        if (startWord <= 0 || endWord <= 0 || startWord > endWord) {
```
Comment (Trifflenator) [inline src/main/java/StringManipulation.java]: I think these exceptions should be separated out. Currently, the exception throw doesn't differentiate between them. This means that for a user they won't know which part was wrong or how to fix it. Not a major issue, more a QoL improvement.

File: src/main/java/StringManipulation.java [hunk 2]
```diff
@@ -1,32 +1,119 @@
 public class StringManipulation implements StringManipulationInterface {
+    private String myString;
 
     @Override
     public String getString() {
-        return null;
+        // Simple check if the String is null or not
+        if (myString == null){
```
Comment (Trifflenator) [inline src/main/java/StringManipulation.java:7]: Why is this necessary? If myString is null, and you return it, that isn't different from what you have here. The reason to check for null would be to throw an exception, I believe, which we aren't doing. This would cut down the code a little. Altogether not a functional change.

File: src/main/java/StringManipulation.java [hunk 3]
```diff
@@ -1,32 +1,119 @@
...
+        }
+
+        // n must be smaller than the length of the string
+        if (n > myString.length()) {
+            throw new IndexOutOfBoundsException("'n' is greater than the string length.");
+        }
+
+        //initial an empty string
+        String result = "";
+        int position = 1;
```
Comment (Trifflenator) [inline src/main/java/StringManipulation.java:47]: This variable doesn't feel necessary. It's used in one place and could be replaced with i+1.

File: src/test/java/StringManipulationTest.java [hunk 4]
```diff
@@ -22,65 +22,88 @@ public void tearDown() {
     @Test
```
Comment (Trifflenator) [inline src/test/java/StringManipulationTest.java:22]: This is a general comment for all the tests:
I've noticed that you aren't testing for exceptions here. I'm not sure if that's a conscious choice or not on your part, so I won't say you need to. I just think that it's better coding practice to test for your exceptions, make sure that they're being thrown when you want them to be. It would also help with some of the tests being similar to each other.

File: src/main/java/StringManipulation.java [hunk 5]
```diff
@@ -1,32 +1,119 @@
...
+        return subString;
     }
 
     @Override
-    public String restoreString(int[] indices) {
-        return null;
+    public String restoreString(int[] indices){
+        //throw an error if user put more indices values than the String length
+        if (indices.length != myString.length()) {
+            throw new IllegalArgumentException("Length of 'indices' must be the same as the string length.");
```
Comment (Trifflenator) [inline src/main/java/StringManipulation.java]: Very minor issue of phrasing, I think saying "Number of indices" instead of  "Length of 'indices'" is a bit clearer to the user.

**Review by Trifflenator — APPROVED** [review]
Overall, I would say that this code is very well written. It implements the methods simply and effectively. I did not notice any out-and-out errors. Most of the comments I had were about minor things that don't affect performance, like an unnecessary variable or if statement. When I ran the code myself, it worked perfectly and passed all included unit tests. As such, I am going to approve the merge, as the changes suggested aren't substantive enough to warrant requesting a change.

**Comment by Monica20030707** [comment]
Thank you for your detail explanation. I have changed 4/5 suggests from you (except for the test cases since I think it is what our Prof might want to see too), but all your suggests for class method change is being updated. Thanks!!
